### DISCUSSION TITLE
Allow .spec.postgresqlVersion in Cluster definition

### DISCUSSION THREAD

**Discussion by aran** [cloudnative-pg/cloudnative-pg#1283] [Ideas]
Currently cloudnative-pg uses a clever convention to retrieve the postgres major version from its own image tag.

As a feature request, allow overriding this convention with an explicit configuration setting, for example, "postgresqlVersion". If specified, this setting would be used by validation code and other logic in place of the tag.

The motivation is a tool called [Skaffold](https://skaffold.dev/). For rapid iteration in dev mode, skaffold generates an image digest to use as a tag and pushes it directly to docker daemon. The image digest ensures that there is never an inconsistency where the kubernetes has the wrong version of an image. The cnpg webhook validation rejects these images.

**Comment by sxd** [comment]
Hi @aran 

Yes, this is used to avoid using images 15 in a cluster that was created with version 14, that will produce a lots of issues, and it's not a minor problem that can be solved easily, major upgrades in PostgreSQL can come with lots of problems to solve, imagine if someone just put a different version (greater or lower) in the storage, that will be a big mess.

Do you have an idea on how to solve that issue? someone putting a version greater or lower than the one in the storage?

Turning this into a discussion

Best Regards!

**Reply by aran** [reply]
Hi @sxd,

Thanks for starting the discussion. The convention-over-configuration paradigm is great and in this case it is being used to enforce an important safety principle around major versions. In this case the feature request is to allow an explicit configuration to override the convention. That leaves the open question of how to recover the safety principle. If I am understanding you right, retaining this safety principle the issue that you're wanting to discuss. Please let me know if I misunderstood.

One way might be to abandon the safety principle for users of the explicit configuration. In this world, the convention keeps data safe by default, but using the new `postgresqlVersion` (or perhaps `postgresqlVersionUnsafe`) configuration setting is in effect an "I know what I'm doing" declaration that waives the safety benefit. Then it becomes my responsibility to ensure I specify the right postgresqlVersion when I upgrade my postgresql container image.

Another option would be to insist on keeping the safety principle in all circumstances. I am not sure the best way to do this from a technical perspective. As you pointed out, HA major version upgrades are a deep topic. One speculative way might to keep the responsibility for validation in the admission controller but modify it to e.g. spin up a Job that runs `postgres -V` with the provided container image. If this is possible, a mutating admission controller could then set `postgresqlVersion` if unset or throw an error in case of mismatch. I don't have experience with admission controllers to know what's possible. 

There's a huge design space available here. I don't know enough about how postgres itself manages major HA version upgrades or the design goals of the operator.

**Reply by aran** [reply]
Can you share any details about how major version upgrades will work? Maybe there's a way to align both requirements.
